README.md
[![Build Status](https://travis-ci.org/metagridch/metagrid-php-client.svg?branch=master)](https://travis-ci.org/metagridch/metagrid-php-client)

This is a small client to use the metagrid API with php.

Metagrid Client Features
======
- Easy to use interface to interact with the metagrid API
- Possibility to change the process of requesting data from the API

Installation
------------
The easiest way to install Metagrid-Client is with [git](http://git.org/).
```bash
git clone https://github.com/metagridch/metagrid-php-client.git
```

Use [composer](http://getcomposer.org) to install dependencies
```bash
php composer.pear install
```

Usage
-----
This is a very simple example.

First create the client. Get all links for provider Dodis with identifier 5
```php
use \Metagrid\MetagridClient;

$client = new MetagridClient();
$links = $client->get('dodis','5')
```
Get all links for a specific resourceType
```php
$links = $client->get('dodis','5','person');
```
Get all links in a specific language
```php
$links = $client->get('dodis','5','person','de');
```
Get all links with more detail information
```php
$links = $client->get('dodis','5','person','de',true);
```
The client use Guzzle as a default http-client library. If you wanna use another http-client just inject it in the constructor. The http-client needs to implements ClientInterface
```php
use \Metagrid\MetagridClient;
use \GuzzleHttp\Client;
$guzzleClient = new Client()
$client = new MetagridClient($guzzleClient);
$links = $client->get('dodis','5')
```
If you like to change the way request to the API are done you can implement your own LinkBuilder. That may be interesting, if you need to double check properties or need to modifies the identifier

```php
class ExampleLinkBuilder implements LinkBuilderInterface {

    const BASEURL = "https://api.metagrid.ch/widget";
    const SEPARATOR = "/";
    // change the defaults slug
    public function make($slug = 'example', $identifier = '', $resourceType = 'person', $lang = 'de', $includeDescription = false)
    {

        //f.e.https://api.metagrid.ch/widget/example/person/5modification.json?lang=de&include=false&jsoncallback=jQuery19109290815709965452_1458038374939&_=1458038374940
        if($this->isExclude($identifier)) return false;
        $result =   self::BASEURL
                    .self::SEPARATOR.$slug
                    .self::SEPARATOR.$resourceType
                    .self::SEPARATOR.$identifier."modification"
                    .".json?lang=".$lang
                    ."&jsoncallback=jQuery".rand(10000,99999)."_".rand(10000,99999);
        if($includeDescription){
            $result .= "&include=true";
        }
        return $result;
    }
    
    /**
    * Exclude resource with identifier 1,2,3
    **/
    private function isExclude($identifier){
        return inArray($identifier, array(1,2,3));
    }
}```



CHANGELOG
CHANGELOG:
---------
0.2
- Updated dependencies
- Add gitlab-ci config file
0.2.1
- Improve documentation
- Improve readme
- Transfer to from gitlab to github

tests/Metagrid/MetagridClientTest.php
<?php

namespace Metagrid\Tests;
require_once(__DIR__ . '/../../vendor/autoload.php');

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use Metagrid\MetagridClient;

/**
 * Created by PhpStorm.
 * User: tobinski
 * Date: 15.03.16
 * Time: 11:43
 */
class MetagridClientTest extends \PHPUnit_Framework_TestCase
{
    /**
     * true = use mocked answers for test
     * false = send request to the metgrid server
     */
    const USE_MOCK = true;

    public function testGet() {
        $client = new MetagridClient($this->getClient($this->shortResponse()));
        $links = $client->get('dds','5');
        // count resources
        $this->assertCount(4, $links);
        // check links (they can change. Its a live check
        $this->assertSame('http://www.hls-dhs-dss.ch/textes/d/D4647.php', $links['HLS']);
        $this->assertSame('http://d-nb.info/gnd/124769942', $links['GND']);
        $this->assertSame('http://www2.unil.ch/elitessuisses/index.php?page=detailPerso&idIdentite=50020', $links['Elites suisses au XXe s.']);
        $this->assertSame('http://resources.huygens.knaw.nl/europeseintegratie/en/persoon/5148', $links['Huygens ING']);
    }

    /**
     * Return descriptions for the widgets. The language of the description can change.
     */
    public function testGetExtended() {
        $client = new MetagridClient($this->getClient($this->extendedResponse()));
        $links = $client->get('dds','5','person','de', true);
        // count resources
        $this->assertCount(4, $links);

        // check links (they can change. Its a live check
        $this->assertSame('http://www.hls-dhs-dss.ch/textes/d/D4647.php', $links['HLS']['url']);
        // check descriptions
        $this->assertArrayHasKey('short_description', $links['HLS']);
        $this->assertArrayHasKey('long_description', $links['HLS']);
        // check urls
        $this->assertSame('http://d-nb.info/gnd/124769942', $links['GND']['url']);
        $this->assertSame('http://www2.unil.ch/elitessuisses/index.php?page=detailPerso&idIdentite=50020', $links['Elites suisses au XXe s.']['url']);
        $this->assertSame('http://resources.huygens.knaw.nl/europeseintegratie/en/persoon/5148', $links['Huygens ING']['url']);
    }

    /**
     * Check the returned values are in the right language.
     * At the moment just the description change. In the next
     * release the link will be language agnostic
     */
    public function testGetLang() {
        $client = new MetagridClient($this->getClient($this->frenchResponse()));
        $links = $client->get('dds','5','person','fr');

        // count resources
        $this->assertCount(4, $links);

        // check if the slug is in the right language
        $this->assertSame('http://www.hls-dhs-dss.ch/textes/d/D4647.php', $links['DHS']);

        // check if FR link is returned. Will be implemented in the next release
        // $this->assertSame('http://www.hls-dhs-dss.ch/textes/f/D4647.php', $links['DHS']['url']);
    }

    /**
     * Get a inexistent resource
     */
    public function testFailure () {
        $client = new MetagridClient($this->getClient('', 404));
        $links = $client->get('dds','111111111111111115','person','de', true);
        $this->assertFalse($links);
    }

    /**
     * Get directly over an url
     */
    public function testGetUrl(){
        $client = new MetagridClient($this->getClient($this->shortResponse()));
        $links = $client->getUrl('https://api.metagrid.ch/widget/dds/person/5.json?jsoncallback=jQuery_2q345');
        $this->assertCount(4,$links);
    }
    /**
     * Wrong URL with exception
     */
    public function testGetWrongUrl(){
        $client = new MetagridClient();
        $this->setExpectedException('InvalidArgumentException');
        $client->getUrl('https://google.ch');
    }

    /**
     * Mock guzzle client
     * @return Client
     */
    private function getClient($content = '', $statusCode = 200) {
        if(self::USE_MOCK) {
            // Mock answers
            $stream = \GuzzleHttp\Psr7\stream_for($content);
            $response[] = new Response($statusCode, [], $stream);

            $mock = new MockHandler($response);

            $handler = HandlerStack::create($mock);
            return new Client(['handler' => $handler]);
        }
        // get from live system
        return new Client();
    }

    /**
     * Mock short answer of the server
     * @return string
     */
    private function shortResponse() {
        // Create a mock and queue two responses.
        $content = array(array(  'HLS' => 'http://www.hls-dhs-dss.ch/textes/d/D4647.php',
            'GND' => 'http://d-nb.info/gnd/124769942',
            'Elites suisses au XXe s.' => 'http://www2.unil.ch/elitessuisses/index.php?page=detailPerso&idIdentite=50020',
            'Huygens ING' => 'http://resources.huygens.knaw.nl/europeseintegratie/en/persoon/5148'));
        return json_encode($content);
    }


    /**
     * Mock wrong answer of the server
     * @return string
     */
    private function frenchResponse() {
        // Create a mock and queue two responses.
        $content = array(array(  'DHS' => 'http://www.hls-dhs-dss.ch/textes/d/D4647.php',
            'GND' => 'http://d-nb.info/gnd/124769942',
            'Elites suisses au XXe s.' => 'http://www2.unil.ch/elitessuisses/index.php?page=detailPerso&idIdentite=50020',
            'Huygens ING' => 'http://resources.huygens.knaw.nl/europeseintegratie/en/persoon/5148'));
        return json_encode($content);
    }


    /**
     * Mock short answer of the server
     * @return string
     */
    private function extendedResponse() {
        // Create a mock and queue two responses.
        $content = array(array(  'HLS' => array(
                'url' => 'http://www.hls-dhs-dss.ch/textes/d/D4647.php',
                'short_description' => 'Historisches Lexikon der Schweiz',
                'long_description' => 'HDas Historische Lexikon der Schweiz sammelt daqten ....'),
            'GND' => array(
                'url' => 'http://d-nb.info/gnd/124769942',
                'short_description' => 'Diplomatische Dokumente der Schweiz',
                'long_description' => 'Die Dodis ...'),

            'Elites suisses au XXe s.' => array(
                'url' =>  'http://www2.unil.ch/elitessuisses/index.php?page=detailPerso&idIdentite=50020',
                'short_description' => 'Elite suisse',
                'long_description' => 'Das Eltie suisse projekt...'),
            'Huygens ING'  => array(
                'url' =>   'http://resources.huygens.knaw.nl/europeseintegratie/en/persoon/5148',
                'short_description' => 'Huygens',
                'long_description' => 'Huygens ist ein projekt ...'))
        );
        return json_encode($content);
    }

}

src/Metagrid/MetagridClient.php
<?php
namespace Metagrid;
require_once(__DIR__ . '/../../vendor/autoload.php');

use \GuzzleHttp\Client;
use \GuzzleHttp\Exception\ClientException;
use \GuzzleHttp\ClientInterface;
use \GuzzleHttp\Exception\GuzzleException;

/**
 * Created by PhpStorm.
 * User: tobinski
 * Date: 15.03.16
 * Time: 10:47
 */
class MetagridClient
{
    /**
     * @var \GuzzleHttp\ClientInterface
     */
    private $client = null;

    /**
     * @var LinkBuilderInterface | null
     */
    private $linkBuilder = null;

    /**
     * MetagridClient constructor.
     * @param LinkBuilderInterface $linkBuilder
     * @param ClientInterface $client
     */
    function __construct(ClientInterface $client = null, LinkBuilderInterface $linkBuilder = null)
    {
        $this->client = $client;
        $this->linkBuilder = $linkBuilder;

    }

    /**
     * Get all links for an identifier and a language
     * @param string $slug Slug of the provider. Usually this is the domain
     * @param string $identifier Identifier of the resource
     * @param string $resourceType ResourceType of the resource. At the moment just person|place|organization
     * @param string $lang Language of the resource
     * @param boolean $includeDescription Include a longer description of the providers in the response
     * @return array|boolean
     */
    public function get($slug = 'dds', $identifier = '', $resourceType = 'person', $lang = 'de', $includeDescription = false){
        $url = $this->getLinkBuilder()->make($slug, $identifier, $resourceType, $lang, $includeDescription);
        if($url === false) return array();
        return $this->getUrl($url);
    }

    /**
     * Get an url directly
     * @param string $url
     * @throws \InvalidArgumentException
     * @return array|bool
     */
    public function getUrl($url = '') {
        if(!preg_match('/api.metagrid.ch/',$url)){
            throw new \InvalidArgumentException("You can just request the metagrid server");
        }
        try {
            $response = $this->getClient()->request('GET', $url, array());
        }
        catch(ClientException $e){
            // todo: log error
            return false;
        }
        catch(GuzzleException $e) {
            // todo: log error
            return false;
        }

        // decode jsonp. In future you can get pure json from the server
        return $this->jsonp_decode($response->getBody()->__toString());

    }

    /**
     * Return a client, Lazy load guzzle Client if no client is configured
     * @return \GuzzleHttp\ClientInterface
     */
    public function getClient () {
        if($this->client == null){
            $this->client = new Client();
        }
        return $this->client;
    }

    /**
     * Set a Client for HTTP request
     * @param \GuzzleHttp\ClientInterface $client
     */
    public function setClient(ClientInterface $client){
        $this->client = $client;
    }


    /**
     * Delegate link generation to an external function
     * @return LinkBuilderInterface
     */
    private function getLinkBuilder(){
        if($this->linkBuilder == null){
            $this->linkBuilder = new DefaultLinkBuilder();
        }
        return $this->linkBuilder;
    }

    /**
     * Decode JSONP to an array.
     * @param  string    $jsonp A jsonp string
     * @return array         an array of the jsonp input
     */
    function jsonp_decode($jsonp = '') {
        if($jsonp[0] !== '[' && $jsonp[0] !== '{')
        {
            $jsonp = substr($jsonp, strpos($jsonp, '('));
        }
        return json_decode(trim($jsonp,'();'), true)[0];
    }
}

src/Metagrid/DefaultLinkBuilder.php
<?php
namespace Metagrid;

/**
 * Created by PhpStorm.
 * User: tobinski
 * Date: 15.03.16
 * Time: 11:31
 */

class DefaultLinkBuilder implements LinkBuilderInterface {

    /**
     * The url of the widget
     */
    const BASEURL = "https://api.metagrid.ch/widget";

    /**
     * The path separator
     */
    const SEPARATOR = "/";

    /**
     * @inheritdoc
     * @param string $slug
     * @param string $identifier
     * @param string $resourceType
     * @param string $lang
     * @param bool $includeDescription
     * @return false|string
     */
    public function make($slug = '', $identifier = '', $resourceType = 'person', $lang = 'de', $includeDescription = false)
    {
        //f.e.https://api.metagrid.ch/widget/dds/person/5.json?lang=en&include=true&jsoncallback=jQuery19109290815709965452_1458038374939&_=1458038374940
        $result =   self::BASEURL
                    .self::SEPARATOR.$slug
                    .self::SEPARATOR.$resourceType
                    .self::SEPARATOR.$identifier
                    .".json?lang=".$lang
                    ."&jsoncallback=jQuery".rand(10000,99999)."_".rand(10000,99999);

        // check if we need to include groups
        if($includeDescription){
            $result .= "&include=true";
        }
        return $result;
    }
}